### PR TITLE
implemented CRubbish, CShinyTexts, CBrightLights, CMotionBlurStreaks, CSpecialFX

### PR DESCRIPTION
**Pull request by aap**

### CODE REVIEW

File: src/render/Rubbish.cpp [hunk 1]
```diff
@@ -1,10 +1,420 @@
...
+		RwRenderStateSet(rwRENDERSTATETEXTURERASTER, RwTextureGetRaster(gpRubbishTexture[type]));
+
+		TempBufferIndicesStored = 0;
+		TempBufferVerticesStored = 0;
+
+		COneSheet *sheet;
+		for(sheet = &aSheets[type*NUM_RUBBISH_SHEETS / 4];
+		    sheet < &aSheets[(type+1)*NUM_RUBBISH_SHEETS / 4];
+		    sheet++){
+			if(sheet->m_state == 0)
```
Comment (erorcun) [inline src/render/Rubbish.cpp:69]: why not an enum? :/

Comment (aap) [reply src/render/Rubbish.cpp:69]: eh, it's so simple...i didn't bother.

File: src/render/SpecialFX.h [hunk 2]
```diff
@@ -87,42 +87,73 @@ class C3dMarkers
...
+	float m_camDist;
+	uint8 m_type;
+	uint8 m_red;
+	uint8 m_green;
+	uint8 m_blue;
+};
+
+class CShinyTexts
+{
+	static int NumShinyTexts;
```
Comment (Sergeanur) [inline src/render/SpecialFX.h:118]: int32, cpp same

**Comment by erorcun** [comment]
CShinyText lgtm

File: src/render/SpecialFX.cpp [hunk 3]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
 {
 }
 
+
+#define BRIGHTLIGHTS_MAX_DIST (60.0f)	// invisible beyond this
+#define BRIGHTLIGHTS_FADE_DIST (45.0f)	// strongest between these two
+#define CARLIGHTS_MAX_DIST (30.0f)
+#define CARLIGHTS_FADE_DIST (15.0f)	// 31 for close lights
+
+int CBrightLights::NumBrightLights;
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:411]: also in32

File: src/render/SpecialFX.cpp [hunk 4]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+static float TallCarHeadLightsFront[8] = { 0.08f, 0.08f, -0.08f, -0.08f, 0.08f, 0.08f, -0.08f, -0.08f };
+static float TallCarHeadLightsUp[8] = { 0.2f, 0.2f, 0.2f, 0.2f, -0.2f, -0.2f, -0.2f, -0.2f };
+static float SirenLightsSide[6] = { -0.04f, 0.04f, 0.06f, 0.04f, -0.04f, -0.06f };
+static float SirenLightsUp[6] = { 0.06f, 0.06f, 0.0f, -0.06f, -0.06f, 0.0f };
+static RwImVertexIndex TrafficLightIndices[4*3] = { 0, 1, 5,  1, 2, 3,  1, 3, 4,  1, 4, 5 };
+static RwImVertexIndex CubeIndices[12*3] = {
+	0, 2, 1,  1, 2, 3,  3, 5, 1,  3, 7, 5,
+	2, 7, 3,  2, 6, 7,  4, 0, 1,  4, 1, 5,
+	6, 0, 4,  6, 2, 0,  6, 5, 7,  6, 4, 5
+};
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:464]: can be const

File: src/render/SpecialFX.cpp [hunk 5]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+				a = 31;
+			else if(aBrightLights[i].m_camDist < CARLIGHTS_MAX_DIST)
+				a = 31 + (255-31)*((aBrightLights[i].m_camDist-CARLIGHTS_FADE_DIST)/(CARLIGHTS_MAX_DIST-CARLIGHTS_FADE_DIST));
+		}
+
+		switch(aBrightLights[i].m_type){
+		case BRIGHTLIGHT_TRAFFIC_GREEN:
+		case BRIGHTLIGHT_TRAFFIC_YELLOW:
+		case BRIGHTLIGHT_TRAFFIC_RED:
+			for(j = 0; j < 6; j++){
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:535]: ARRAYSIZE

File: src/render/SpecialFX.cpp [hunk 6]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+			}
+			for(j = 0; j < 4*3; j++)
+				TempBufferRenderIndexList[TempBufferIndicesStored+j] = TrafficLightIndices[j] + TempBufferVerticesStored;
+			TempBufferVerticesStored += 6;
+			TempBufferIndicesStored += 4*3;
+			break;
+
+		case BRIGHTLIGHT_FRONT_LONG:
+		case BRIGHTLIGHT_REAR_LONG:
+			for(j = 0; j < 8; j++){
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:550]: also

File: src/render/SpecialFX.cpp [hunk 7]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+			}
+			for(j = 0; j < 12*3; j++)
+				TempBufferRenderIndexList[TempBufferIndicesStored+j] = TrafficLightIndices[j] + TempBufferVerticesStored;
+			TempBufferVerticesStored += 8;
+			TempBufferIndicesStored += 12*3;
+			break;
+
+		case BRIGHTLIGHT_FRONT_SMALL:
+		case BRIGHTLIGHT_REAR_SMALL:
+			for(j = 0; j < 8; j++){
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:566]: also

File: src/render/SpecialFX.cpp [hunk 8]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+			}
+			for(j = 0; j < 12*3; j++)
+				TempBufferRenderIndexList[TempBufferIndicesStored+j] = TrafficLightIndices[j] + TempBufferVerticesStored;
+			TempBufferVerticesStored += 8;
+			TempBufferIndicesStored += 12*3;
+			break;
+
+		case BRIGHTLIGHT_FRONT_TALL:
+		case BRIGHTLIGHT_REAR_TALL:
+			for(j = 0; j < 8; j++){
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:582]: also

File: src/render/SpecialFX.cpp [hunk 9]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+				RwIm3DVertexSetPos(&TempBufferRenderVertices[TempBufferVerticesStored+j], pos.x, pos.y, pos.z);
+			}
+			for(j = 0; j < 12*3; j++)
+				TempBufferRenderIndexList[TempBufferIndicesStored+j] = TrafficLightIndices[j] + TempBufferVerticesStored;
+			TempBufferVerticesStored += 8;
+			TempBufferIndicesStored += 12*3;
+			break;
+
+		case BRIGHTLIGHT_SIREN:
+			for(j = 0; j < 6; j++){
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:597]: also

File: src/render/SpecialFX.cpp [hunk 10]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+		case BRIGHTLIGHT_TRAFFIC_YELLOW:
+		case BRIGHTLIGHT_TRAFFIC_RED:
+			for(j = 0; j < 6; j++){
+				pos = TrafficLightsSide[j]*aBrightLights[i].m_side +
+					TrafficLightsUp[j]*aBrightLights[i].m_up +
+					aBrightLights[i].m_pos;
+				RwIm3DVertexSetRGBA(&TempBufferRenderVertices[TempBufferVerticesStored+j], r, g, b, a);
+				RwIm3DVertexSetPos(&TempBufferRenderVertices[TempBufferVerticesStored+j], pos.x, pos.y, pos.z);
+			}
+			for(j = 0; j < 4*3; j++)
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:542]: ARRAYSIZE(TrafficLightIndices)

File: src/render/SpecialFX.cpp [hunk 11]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+				TempBufferRenderIndexList[TempBufferIndicesStored+j] = TrafficLightIndices[j] + TempBufferVerticesStored;
+			TempBufferVerticesStored += 6;
+			TempBufferIndicesStored += 4*3;
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:545]: for these parts it could be also used

File: src/render/SpecialFX.cpp [hunk 12]
```diff
@@ -402,6 +402,379 @@ C3dMarkers::Update()
...
+			for(j = 0; j < 8; j++){
+				pos = LongCarHeadLightsSide[j]*aBrightLights[i].m_side +
+					LongCarHeadLightsUp[j]*aBrightLights[i].m_up +
+					LongCarHeadLightsFront[j]*aBrightLights[i].m_front +
+					aBrightLights[i].m_pos;
+				RwIm3DVertexSetRGBA(&TempBufferRenderVertices[TempBufferVerticesStored+j], r, g, b, a);
+				RwIm3DVertexSetPos(&TempBufferRenderVertices[TempBufferVerticesStored+j], pos.x, pos.y, pos.z);
+			}
+			for(j = 0; j < 12*3; j++)
+				TempBufferRenderIndexList[TempBufferIndicesStored+j] = TrafficLightIndices[j] + TempBufferVerticesStored;
```
Comment (ShFil119) [inline src/render/SpecialFX.cpp:559]: out of array scope? `TrafficLightIndices` has size 4*3

File: src/render/SpecialFX.h [hunk 13]
```diff
@@ -87,42 +81,125 @@ class C3dMarkers
...
+	float m_camDist;
+	uint8 m_type;
+	uint8 m_red;
+	uint8 m_green;
+	uint8 m_blue;
+};
+
+class CBrightLights
+{
+	static int NumBrightLights;
```
Comment (ShFil119) [inline src/render/SpecialFX.h:133]: int32

File: src/render/Rubbish.cpp [hunk 14]
```diff
@@ -1,10 +1,420 @@
...
+void
+CRubbish::Render(void)
+{
+	int type;
+
+	RwRenderStateSet(rwRENDERSTATEZWRITEENABLE, (void*)FALSE);
+	RwRenderStateSet(rwRENDERSTATEVERTEXALPHAENABLE, (void*)TRUE);
+	RwRenderStateSet(rwRENDERSTATEFOGENABLE, (void*)TRUE);
+
+	for(type = 0; type < 4; type++){
```
[the remark this answers is not in the document]
Comment (aap) [reply src/render/Rubbish.cpp:59]: It is, this is for the 4 types (only texture is different in III)

**Pull request merged**